src/main/java/org/zalando/nakadi/validation/JsonSchemaLoader.java

```diff
@@ -0,0 +1,191 @@
+package org.zalando.nakadi.validation;
+
+import com.google.common.collect.Lists;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.zalando.nakadi.domain.CompatibilityMode;
+import org.zalando.nakadi.domain.EventType;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+public class JsonSchemaLoader {
+    public static final String DATA_CHANGE_WRAP_FIELD = "data";
+
+    private static final String ADDITIONAL_PROPERTIES = "additionalProperties";
+    private static final String ADDITIONAL_ITEMS = "additionalItems";
+    private final List<String> keywordsWithNestedSchemas = Lists.newArrayList("definitions", "dependencies",
+            "properties");
+    private final List<String> keywordsInferObjectSchema = Lists.newArrayList("required", "minProperties",
+            "maxProperties");
+    private final List<String> keywordsInferArraySchema = Lists.newArrayList("minItems", "maxItems", "uniqueItems");
```

Comment:
Sometimes, schema objects don't have explicit types. So we need to infer based on some properties. Everit library does the same https://github.com/everit-org/json-schema/blob/36b7d3f407b32066f380e91e91a2dc6e84ac66ff/core/src/main/java/org/everit/json/schema/loader/SchemaLoader.java#L349-L359